jobs/retry_create.go

```diff
@@ -0,0 +1,137 @@
+package jobs
+
+import (
+	"time"
+
+	"github.com/RedHatInsights/sources-api-go/dao"
+	l "github.com/RedHatInsights/sources-api-go/logger"
+	m "github.com/RedHatInsights/sources-api-go/model"
+	"github.com/RedHatInsights/sources-api-go/service"
+	"github.com/RedHatInsights/sources-api-go/util"
+	"gorm.io/gorm"
+)
+
+const (
+	RETRY_MAX        = 5
+	RECORD_AGE_LIMIT = -1 * 30 * time.Minute
+)
+
+type RetryCreateJob struct{}
+
+// implementing the interface - but these functions aren't really needed since
+// this is a scheduled job.
+func (r *RetryCreateJob) Delay() time.Duration              { return 0 }
+func (r *RetryCreateJob) Arguments() map[string]interface{} { return map[string]interface{}{} }
+func (r *RetryCreateJob) Name() string                      { return "RetryCreateJob" }
+func (r *RetryCreateJob) ToJSON() []byte                    { panic("not implemented") }
+
+// run the job, using any args on the struct
+func (r *RetryCreateJob) Run() error {
+	// running all of this as a transaction so it is idempotent if something goes wrong.
+	return dao.DB.Transaction(func(tx *gorm.DB) error {
+		// find all applications with retry counter > 5 and available, update
+		// retry counter to 5 so they don't get picked up again.
+		result := tx.Debug().
+			Model(&m.Application{}).
+			Where("availability_status = ? AND retry_counter < ?", m.Available, RETRY_MAX).
+			Update("retry_counter", RETRY_MAX)
+		if result.Error != nil {
+			l.Log.Errorf("Error updating available applications' retry counters")
+			return result.Error
+		}
+
+		l.Log.Infof("Updated %v applications that became available since last run but had less retry counters", result.RowsAffected)
+
+		// find all applications that are unavailable/null/empty
+		// AND
+		// created_at less than 30m ago
+		// AND
+		// retry counter less than configured amount
+		apps := make([]m.Application, 0)
+		result = tx.Debug().
+			Select("id", "tenant_id", "application_type_id").
+			Model(&m.Application{}).
+			Where("availability_status != ? ", m.Available).
```

Comment:
it did not take when records with availability_status are null.
We can use `IS DISTINCT FROM`.
```
SELECT id,tenant_id,application_type_id, availability_status FROM applications WHERE (availability_status IS DISTINCT FROM 'available' ) AND created_at > '2022-06-01 15:15:10';
```